### ISSUE TITLE
LFX Workspace : Create an LLM app with deep understanding of a GitHub repo

### ISSUE THREAD

**Issue by staru09** [WasmEdge/WasmEdge#3725]
### Project Title

An LLM app with a deeper understanding of a GitHub repo

### Motivation

It becomes challenging to review PRs and solve issues for huge repos. So the primary motivation for building this project is to automate processes like PR and issue review smoothly.

### Expected Outcome

An LLM bot that can provide accurate comments and code suggestions for new PRs or issues as needed.

### Details

- The bot will take the repo's content as context before commenting.
- It will review the new issues and PRs taking the above context into account.
- Using this context, the bot will provide accurate comments or code suggestions without hallucinating.

### Milestones

- [ ] (M1) [1 week] : Get a brief understanding of the project and discuss milestones
- [ ] (M2) [3 week] : Experiment with various methods to analyze the repository
- [ ] (M3) [3 week] : Create a knowledge base with the contents of the repository
- [ ] (M4) [2 week]: Test this knowledge base with existing bots or create a new bot for the task
- [ ] (M5) [2 week]: Improving the bot's response and optimizing the workflow
- [ ] (M6) [1 week]: Working on documentation

These milestones are flexible and bound to change as the project proceeds.

### Appendix

#3581 
[Gaianet](https://docs.gaianet.ai/category/agent-frameworks-and-apps)
[Flows_network](https://docs.flows.network/docs/intro)

**Comment by staru09** [comment]
Greetings@juntao
Kindly review this and suggest anything that needs to be added/ changed.

**Comment by juntao** [comment]
Thanks. @staru09 I think we have two goals for the mentorship.

1 Develop a good QA system for the repo. That is to allow users to ask questions about the repo and get reasonable answers. Furthermore, the user should be able to ask for a new feature and get the LLM to output code for a specific file in the repo. For this, I believe we will need to build a vector knowledge base for the repo. We could

1a. Generate a summary + 10 QAs for each file in the repo
1b. Create a vector for each generated summary / QA. The payload of the vector should be content of the file.
1c. Build a Gaia node with this knowledge base. When the user asks a question, it will search the vectors to find the right file and then put the file conteent (vector payload) into the system prompt

You can learn more about this approach including tools we developed here:

https://huggingface.co/datasets/gaianet/learn-rust

2 Develop a series of "code review" prompts. We already knew that we should put all files in a PR into the "system prompt" and then ask the LLM to review a specific file. But what questions should we ask for the review? For example, this article discusses 6 common mistakes in Rust:

https://medium.com/@tzutoo/rust-common-mistakes-8e759c6e1dc

Perhaps we can ask the LLM 6 times. Each time with a mistake and its patterns in the prompt, and ask the LLM to identify whether the source code contains this pattern, and suggest how to rewrite the code to fix it.

You can use the public Gaia nodes, such as `llama`, `phi`, and `codestral` to do this research.

https://docs.gaianet.ai/user-guide/nodes

Please keep us updated with your progress. Thanks.

**Comment by staru09** [comment]
Sure sir right now I am working on how to parse various files inside a repo.. 

1.  While creating a knowledge base from a repo should I use only the files containing code or all the files inside the repo? 
2. Files containing code and files like an image or a PDF will be handled separately. ( using OCR maybe)
3.  Different ways to summarize and generate Q/As will be used for these file types.

**Comment by juntao** [comment]
Sounds good. I think we want all files but the non-text files are much less
important (can be ignored for now) since we cannot put them into a prompt
for the LLM to understand.

Cheers
Michael

**Comment by staru09** [comment]
1. I have written this [script](https://github.com/staru09/Github_analyser/blob/main/scripts/Data_fetch/3.%20Code_parser.py) that reads the files with a specific extension and saves the content in a [csv](https://github.com/staru09/Github_analyser/blob/main/Output/repo_contents.csv) file.
2. The next step was to summarize that CSV file and generate ques and ans from the code we added from the repo.
3. I used [this](https://github.com/staru09/Github_analyser/blob/main/scripts/Summary/code_summarizer.py) to summarize and [this](https://github.com/staru09/Github_analyser/blob/main/scripts/Summary/qna_generate.py) to generate question and answers using different models like llama, gemma and phi. Here are the [results](https://github.com/staru09/Github_analyser/tree/main/Output).

4. The next step is to combine multiple LLMs for summarization and ques answer generation. One way to do it is to summarize using one LLM and generate questions using another. Another way to do it can be summarizing separately using these LLMs and merging them later by a different LLM altogether. 

Kindly let me know if anything else could be done or if I missed something.

**Comment by staru09** [comment]
1. I have been able to extract the data and generate a csv from all the links mentioned above. They can be found [here](https://github.com/staru09/Github_analyser/tree/main/Output/main_repos).
2.  I have also extracted and summarized the csv on a smaller [repo](https://github.com/staru09/LFX_test) in the format suggested above and [this](https://github.com/staru09/Github_analyser/blob/main/Output/summary/repo_codes_split.csv) is the result.
3.  I deployed a llama 3.1 node using this knowledge base and these are the respones.
[Screencast from 2024-09-12 20-50-25.webm](https://github.com/user-attachments/assets/da86f75c-3fae-4e0b-bbd1-f831994b4f58)

4. The next task will be to generate similar knowledge bases from the above mentioned data sources then build a node and then test it.

**Comment by juntao** [comment]
These look good. I would like to do one repo first and then test some application scenarios:

1 Ask questions related to the repo in the chatbot UI. Note: use a system prompt like this: 

https://github.com/GaiaNet-AI/node-configs/blob/main/llama-3.1-8b-instruct_rustlang/config.json#L20

Source: https://x.com/mckaywrigley/status/1833941986711396645

2 As it it to implement some new feature for the repo.

3 Connect it to Cursor or Zed as the backend LLM (basically the same task as above but with a new UI)

4 Use it to review PRs (You can use the chatbot UI and put the PR content in the prompt to test)

**Comment by staru09** [comment]
Sure sir

**Comment by staru09** [comment]
Greetings sir 
1. I was trying to summarize the csv files that I have created from the [wasmedge_repo](https://github.com/WasmEdge/WasmEdge).
I was able to create [this](https://github.com/staru09/Github_analyser/blob/main/Output/split_summary/wasmedge_all_split.csv) and it follows the same patter that you had suggested earlier but it is just a small part of the complete csv. I tried running the [script](https://github.com/staru09/Github_analyser/blob/main/scripts/Summary/final_split_summary.py) on the whole file but got this error (llama node is down or something like that)

openai.NotFoundError: <!DOCTYPE html>

      <div class="info">
        <h2>GaiaNet node is not started</h2>
        <p>
          Visit
          <a href="https://www.gaianet.ai" target="_blank">https://www.gaianet.ai</a>
          for more information
        </p>
      </div>
    </div>
  </body>
</html>

from whatever files I was able to summarize and generate qna, here is a [snapshot](https://huggingface.co/datasets/FiendHunter/Github_bot/resolve/main/default-7399382323022328-2024-09-23-00-35-27.snapshot)  and these are the results with chatbot UI using the snapshot 
[Screencast from 2024-09-23 06-56-44.webm](https://github.com/user-attachments/assets/29937415-e75d-4500-8f58-7ec831ae9145)


I am currently trying to summarize using gemma and will update with the results here itself.

2. I was able to summarize and generate qna from markdown files of [ethereum](https://github.com/ethereum/ERCs) repo but while I was creating a snapshot I got this error. 

![image](https://github.com/user-attachments/assets/83938e59-b97e-4384-8cbb-25b2d51a1ae0)

I tried changing vector size to 1000 and context_size to 16000 but then got a different error 
![image](https://github.com/user-attachments/assets/0560fda0-65ee-4a8a-b828-e32b87e6d3f6)

I even tried changing the [main.rs](https://github.com/GaiaNet-AI/embedding-tools/blob/main/csv_embed/src/main.rs) but I was unable to build the wasm app as given [here](https://github.com/GaiaNet-AI/embedding-tools/tree/main/csv_embed) using these commands

`RUSTFLAGS="--cfg wasmedge --cfg tokio_unstable" cargo build --target wasm32-wasi --release
cp target/wasm32-wasi/release/csv_embed.wasm .`

3. I also tried zed and cursor using the default models and with the api keys you have given and results were pretty good. 

My plan for this week is to
1. Summarize and create snapshots keeping the context size in mind.
2. With those snapshots I will be creating nodes that can be tested with chatbot UI or Zed or Cursor. 

Lastly, I made some changes in the PR review bot and changed the node to yibcoder but it seems to generate no text results. 
![image](https://github.com/user-attachments/assets/97b7ba20-fc86-4c97-9043-b938df88cb46)
I even tried to generate summary using that node but it gave an error saying unable to complete request.

Kindly suggest how I can get past this context size related errors.

**Comment by juntao** [comment]
Thanks.

Re 1. The llama.us.gaianet.network node is sometimes down this week since we have a hackathon in Singapore that made heavy use of this node. I would not mind using GPT 4o to generate high quality Q&As if needed. You can change the Python script to point to openAI models -- let me know if you need an API key for this.

Look at the node log, it also appears that you are trying to generate tool-call responses. I would note that the regular Llama 3.1 model cannot handle tool-calls very well. You need to use this node:

https://docs.gaianet.ai/user-guide/nodes#tool-use-llama

Example: https://docs.gaianet.ai/tutorial/tool-call 

Re 2. The vector dimension is determined by the embedding model. For nomic-embed, it is 768, you cannot change it. But the error you saw is not related to that either -- it means that your input chunk is longer than 8192 tokens. 

That is why we took the CSV approach -- we generate short QAs for long text so that the embedding text is alwasy short even as the original text may exceed 8192 tokens.

If you really want to embed the markdown sections, you will need manually break up the long sections into smaller markdown sections. Each section must be shorter than 8192 tokens (around 5000 words)

**Comment by staru09** [comment]
Greetings sir
1. I was able to generate summary and qna from the full repos of llamaedge and wasmedge, this is how the final CSVs look like  [wasmedge](https://huggingface.co/datasets/FiendHunter/Github_bot/resolve/main/wasmedge_split_clean.csv) , [llamaedge](https://huggingface.co/datasets/FiendHunter/Github_bot/resolve/main/llama_edge_split.csv) and then I created their [snapshots](https://huggingface.co/datasets/FiendHunter/Github_bot/tree/main). 
2. With these snapshots I was able to run llama 3.1_8B_Q5 model using chat
bot UI and [these](https://drive.google.com/file/d/19sbpvsbCkIj77vob6EECzsc7LrkgHJL7/view?usp=sharing) are the results with wasmedge snapshot. Here are the prompts and node specs.
![Screenshot from 2024-09-28 05-11-15](https://github.com/user-attachments/assets/47535da9-5ebe-4c65-88e9-f5c9af871522)
3. However when I tried to chat using zed or cursor as backend I got an error message of incorrect api.
![incorrect_api](https://github.com/user-attachments/assets/ca83c3eb-baf4-4bb3-8b00-57712497cbf9)
I tried running the node both way (normal and local only) but got the error message. I even tried running the node on 1 device and running IDEs on another but still got the same error message. 
Can you please suggest why is this happening or do I need to run the nodes differently.
4. I also wrote a [summarizer](https://github.com/staru09/Github_analyser/blob/main/scripts/Summary/openai_summary.py) script that uses openai api keys but the keys that you have shared with me earlier are not working. Maybe there's a mismatch between model name and the keys that are being used. Please share with me the updated keys so I can test the summarization and if there is any major difference, then we will try to improve llama using better prompts or some other ways.
5. Lastly, regarding the issues that I faced last week while using markdown files I wanted to ask if it will be a good idea to delete all those rows that have a word count greater than 8000 tokens or if not delete, shall I truncate them to the same limits?

Kindly suggest what else can be done.

**Comment by juntao** [comment]
Not sure if I answered No.3 here. The 8192 is the context length. It is the max length of your generated QA pair. 

The actual content that generates the QA could be much longer. They are not turned into vectors. They are stored in the vectors DB as the "payload" for the vector created from the QA pair.

**Comment by juntao** [comment]
Also check out those two videos:

https://youtu.be/icbFAAOZYcE

https://youtu.be/TFgrdIA-4B8

I am thinking about creating a demo like that 

1 create a gaia node with knowledge about repo A (it should be a code example / docs / education repo)

2 use the Gaia node + Zed to create an app for that SDK. 

What do you think?

**Comment by staru09** [comment]
That would be great!!!
I have the knowledge base ready 
I will create snapshots and test them again with chatbot UI.
Once this is done, we can proceed with zed or cursor for creating demo.

Although I wasn't able to run my local node using zed or cursor.

**Comment by juntao** [comment]
OK. Once you have it tested, I will create a public node for you to test in the IDEs! Look forward to it.

**Comment by staru09** [comment]
Greetings sir 
[This](https://huggingface.co/datasets/FiendHunter/Snapshots_github/tree/main) contains all the snapshots.
You can create a node with wasmedge and llamaedge snapshot for now. 
I have tested them with chatbot UI, and there was a little hallucination with Lama 3.1 8b while chatting, but the code it wrote was somehow accurate.

**Comment by juntao** [comment]
Great. Can you upload the intermediary files in the HF repo so that other people can duplicate it?

* The CSV file used for creating vectors
* The script to create the CSV from a collection of files
* The script to generate the file collection from a repo

Example: in the learn-rust repo, we have these assets from the book text to the CSV to the snapshot. 

https://huggingface.co/datasets/gaianet/learn-rust/resolve/main/rust-books.txt

https://huggingface.co/datasets/gaianet/learn-rust/resolve/main/vectors_from_paragraph_rust.py

https://huggingface.co/datasets/gaianet/learn-rust/resolve/main/rust-books-pairs.csv

https://huggingface.co/datasets/gaianet/learn-rust/resolve/main/rust-books.snapshot.tar.gz

**Comment by staru09** [comment]
Sure sir
1. [This](https://github.com/staru09/Github_analyser/blob/main/scripts/Data_fetch/2.%20Repo_parser.py) is the code I used to create a file collection from a repo.
It returns a CSV file like the one given [here.](https://huggingface.co/datasets/FiendHunter/Github_bot/tree/main/main_repos) 


2. After creating the file collection, [this](https://github.com/staru09/Github_analyser/blob/main/scripts/Summary/final_split_summary.py) is the code I used to create summary and generate several qna pairs.
It returns a CSV file like [this.](https://huggingface.co/datasets/FiendHunter/Github_bot/tree/main/split_summary)

3. Sometimes the code gives an error message so I created separate files to summarize and generate qna pairs.
The [summarizer](https://github.com/staru09/Github_analyser/blob/main/scripts/Summary/code_summarizer.py) script generates an intermediary file like [this](https://huggingface.co/datasets/FiendHunter/Github_bot/tree/main/summary) which can be used to generate the knowledge base similar to the one given in the previous step using [this](https://github.com/staru09/Github_analyser/blob/main/scripts/Summary/qna.py) script. (The second approach takes slightly longer than the previous one)


4. Then I followed the steps on the [Gaianet](https://docs.gaianet.ai/creator-guide/knowledge/csv) website to create snapshots from the final knowledge base. All the snapshots can be found [here](https://huggingface.co/datasets/FiendHunter/Snapshots_github).

5. I will update the readme file in the repo to make it easy to replicate the results.

**Appendix**
1. [Main_repo](https://github.com/staru09/Github_analyser)
2. [Snapshots](https://huggingface.co/datasets/FiendHunter/Snapshots_github)

**Comment by juntao** [comment]
Great. Please copy the source files and scripts into that Huggingface repo together with the snapshot files. This way, a new user will be able to find all he needs in a single repo. Thanks. 

Also, have you seen this?

https://github.com/abinthomasonline/repo2txt

It would be nice to have a script that generates QnA CSV files based on the output from this tool.

**Comment by staru09** [comment]
I'll add all the files mentioned above in a hf repo.
I'll also test the above library and share the results.

**Comment by staru09** [comment]
I tested both the tools mentioned above.

1. With regards to [repo2txt](https://github.com/abinthomasonline/repo2txt ) 

- It allows to create a txt file with full contents of a repo and it also allows to select files with specific extension and download the partial repo.
- I basically created [this](https://github.com/staru09/Github_analyser/blob/main/scripts/Data_fetch/local_parser.py) script that converts the partial repo in a csv with Path and Content as headers. This csv can be used to generate qna pairs as we were doing before.
- I created [this](https://github.com/staru09/Github_analyser/blob/main/Output/local_repo/llamaedge.csv) csv file using [llamedge](https://github.com/LlamaEdge/LlamaEdge) repo but instead of choosing all the files I used only .rs, .sh and . md files. 
![repo2txt](https://github.com/user-attachments/assets/d9702261-6c03-4215-9b6b-dffda9583fe0)


2. Now coming to [repopack](https://github.com/yamadashy/repopack)
- [This](https://github.com/staru09/Github_analyser/blob/main/scripts/Data_fetch/repopack_to_csv.py) is the script I used to generate a csv from this [text](https://github.com/staru09/Github_analyser/blob/main/Output/local_repo/llamedge_complete.txt) file. 
- The txt file was generated after cloning the repo locally and using 
`repopack` command inside the cloned repo.
- The generated csv file can be used to generate qna pairs.

**Comment by juntao** [comment]
Hmm, the CSV you generated has two columns -- path and content. I think our CSV tool requires a single column for the content. 

I would suggest to add the path to the file content using the cursor AI format like this

```
``````language:path/to/file
{file content}
``````
```

Then, remove the path column. 

We also need a script to process the text output from repo2txt in addition to the repo folder.

**Comment by juntao** [comment]
> Also, is [this](https://github.com/GaiaNet-AI/embedding-tools/tree/main/csv_embed) the CSV tool that you are referring above?

Yeah, we use the csv_embed.wasm to create vectors from the two-column CSV. 

Do you have another Python script to generate the vector snapshots from the 3-column CSV?

I think adding the path into the content column will help us generate Cursor AI (and other IDEs like Zed) compatible text for the LLM context. 

So, if you are using a different Python script to generate the vectors, we could modify that script to add path to content (in cursor ai format) before adding the vector point. Thanks.

**Comment by staru09** [comment]
> Do you have another Python script to generate the vector snapshots from the 3-column CSV?


No sir I don't have a different python script to generate vectors from a 3-column CSV, even I was using csv_embed.wasm to create vectors. 
I think I can do this by adding Path to the Content column before or after summarization (depending on its impact on summaries and qna pairs).

so the final content will follow this structure i.e. the `Content` column should contain the text in the format given below:-

`Language:path/to/file`           
`{file content}`  

I'll create a python script for the same.

**Comment by juntao** [comment]
OK! Sorry I was not clear before. But yes, the CSV that goes into csv_embed.wasm should only have two columns

1 The content (or the article) -- that is the full content of the file
2 a QnA

I think you can add PATH to the CONTENT before generating the QnA. It is a bit more of information for the QnA generator. 

Also, I think we might do this more intelligently. For example, we can use the following for code files.

    ```
    ```language:path/to/file
    {file content}
    ```

But the following for MD or doc files

    The following a markdown document located at path/to/file
    ------
    {file content}
    ------

**Comment by staru09** [comment]
[This](https://github.com/staru09/Github_analyser/blob/main/scripts/Helper/local_csv.py) is a simple Python script that adds a path to the content column (for code)
[Here](https://github.com/staru09/Github_analyser/blob/main/Output/local_repo/llamaedge_merge.csv) is the output. 
Would you like me to add delimiters also or is this fine?

**Comment by juntao** [comment]
For code files, we should enclose the entire file in the markdown code blocks. And the `language` should be replaced by the language determined by the file name suffix. Here is an example

    ```rust:crates/llama-core/src/error.rs
    //! Error types for the Llama Core library.

    use thiserror::Error;

    /// Error types for the Llama Core library.
    #[derive(Error, Debug)]
    pub enum LlamaCoreError {
        /// Errors in General operation.
        #[error("{0}")]
        Operation(String),
        /// Errors in Context initialization.
        #[error("Failed to initialize computation context. Reason: {0}")]
        InitContext(String),
        /// Errors thrown by the wasi-nn-ggml plugin and runtime.
        #[error("{0}")]
        Backend(#[from] BackendError),
        /// Errors thrown by the Search Backend
        #[cfg(feature = "search")]
        #[error("{0}")]
        Search(String),
    }

    /// Error types for wasi-nn errors.
    #[derive(Error, Debug)]
    pub enum BackendError {
        /// Errors in setting the input tensor.
        #[error("{0}")]
        SetInput(String),
        /// Errors in the model inference.
        #[error("{0}")]
        Compute(String),
        /// Errors in the model inference in the stream mode.
        #[error("{0}")]
        ComputeSingle(String),
        /// Errors in getting the output tensor.
        #[error("{0}")]
        GetOutput(String),
        /// Errors in getting the output tensor in the stream mode.
        #[error("{0}")]
        GetOutputSingle(String),
        /// Errors in cleaning up the computation context in the stream mode.
        #[error("{0}")]
        FinishSingle(String),
    }
    ```

**Comment by staru09** [comment]
Sir this is the final [script](https://github.com/staru09/Github_analyser/blob/main/scripts/Helper/local_csv.py) . It generates [output](https://github.com/staru09/Github_analyser/blob/main/Output/local_repo/llamaedge_merge.csv) in the format suggested above. I'll update the script for other repos depending on the type of files (other extensions like .js,.tsx etc). This one was specifically for llamaedge.

**Comment by juntao** [comment]
When I open the CSV file in excel (or google spreadsheet), it looks like this:

<img width="412" alt="Screenshot 2024-10-08 at 3 21 01 PM" src="https://github.com/user-attachments/assets/209f6f97-8926-45b5-9acb-4d872d402bcb">

I think there are un-escaped commas in the file. You should use the Python CSV library to create the CSV document instead of just combining strings together. 

After that, you can run the script to create QnAs -- make sure that you make changes to the prompt so that it can generate appropriate QnAs for code files.

https://github.com/GaiaNet-AI/embedding-tools/blob/main/csv_embed/vectors_from_csv.py

You could change the LLM in the script to our public Qwen coder node.

```
API_BASE_URL = "https://qwencoder.us.gaianet.network/v1"
MODEL_NAME = "qwencoder"
API_KEY = "GAIA"
```

---

Also, please make the following change:

FROM

```
else:
            new_content = f"language: {path}\n{content}"
```

TO

```
else:
            new_content = f"The following document is located at {path}\n------\n{content}\n------"
```

Thanks.

**Comment by staru09** [comment]
1. Sir, even though I encountered this very issue when I was opening the csv in MS Excel, the output became clear when I opened it in Google Sheets. 
![image](https://github.com/user-attachments/assets/467a571b-e18f-44d2-89d7-a08b995f86e8)
2. I have updated the [script.](https://github.com/staru09/Github_analyser/blob/main/scripts/Helper/local_csv.py)

**Comment by juntao** [comment]
The easy way to test is to run it through the script for QnA generation. If it treats each file as a line, we are good! :) Can you try it?

**Comment by staru09** [comment]
Yes sir here are the [results](https://github.com/staru09/Github_analyser/blob/main/Output/local_repo/llamaedge_merge_split.csv). 
Maybe the above issue was related to how excel interprets a csv or something like that.

**Comment by juntao** [comment]
I do not get it ...

The processed file 

https://github.com/staru09/Github_analyser/blob/main/Output/local_repo/llamaedge_merge_split.csv

only has 73 lines, which correspond to 10 different files (each file has multiple QAs).

The original file

https://github.com/staru09/Github_analyser/blob/main/Output/local_repo/llamaedge_merge.csv

has 60 lines, which correspond to 60 files.

**Comment by staru09** [comment]
Sir, I am very sorry for the confusion. The output file I shared is only a part of the complete output; the complete file is still being processed. I'll update once it's completely processed, the test file was only to test if is the content is being read by the script correctly.

**Comment by juntao** [comment]
Ok. I was asking because the abnormal CSV entries were after row #10. So, in order to prove that the format is correct, we need need to convert all 60 files ...

**Comment by staru09** [comment]
Here is the final [output](https://github.com/staru09/Github_analyser/blob/main/Output/local_repo/llamaedge_short_split.csv) file. I deleted some rows from this that were giving an out of context error by the api endpoints. 
I am creating a snapshot will update here in a while.

**Comment by juntao** [comment]
Thanks. Yes, the output file looks good.

**Comment by staru09** [comment]
I'll create snapshots for other repos as well and update.

**Comment by staru09** [comment]
Good morning sir
I just wanted to confirm which specific file extensions I should retrieve from the main repositories (wasmedge, llamaedge, and quickjs). These repositories contain various files with different extensions, many of which may not be relevant to us (e.g., `.manylinux2014_aarch64`, `.manylinux2014_x86_64`, `.manylinux2014-build-plugins-deps`, `.release`, `.ubuntu-base`, `.ubuntu-cuda`, `.ubuntu-env` and a lot more), from the documentation repository, should I only include `.md` or `.txt` files, or should we also include HTML or other code files?

**Comment by juntao** [comment]
You may need to research for each project. For example, for WasmEdge, MD files under this directory would be great for the documentation knowledge base:

https://github.com/WasmEdge/docs/tree/main/docs

For LlamaEdge:

https://github.com/Llamaedge/docs/tree/main/docs

But for QuickJS, I think the examples are best

https://github.com/second-state/wasmedge-quickjs/tree/main/examples
https://github.com/second-state/wasmedge-quickjs/tree/main/example_js

It is possible that we need to need to hand write some QnAs for those example source code files to indicate what the example is for.

**Comment by staru09** [comment]
Sir, what will be the purpose of handwritten QnAs are we going to fine-tune llama for QnA generation or is it just for testing purposes?

**Comment by staru09** [comment]
Greetings sir,
I created csv in the format required by the Cursor IDE, but certain rows have more than 5000 words, so llama cannot summarize or generate QnA and the code crashed there. One way is to delete those specific rows that have more than 5000 words else we can split the rows but it will result in loss of context I feel. Can you suggest what else can be done?

**Comment by juntao** [comment]
Hmm, 5000 words is just 8k tokens. It should not be a problem for llama 3.1 or 3.2 with 128k context sizes. 

What is the error message you see when you crashed the service with the long text?

**Comment by staru09** [comment]
This is the error message that I get when the text exceeds 5000 words

```Traceback (most recent call last):
  File "c:\Users\91745\OneDrive\Desktop\Github_analyser\scripts\Summary\summarizer.py", line 120, in <module>
    main()
  File "c:\Users\91745\OneDrive\Desktop\Github_analyser\scripts\Summary\summarizer.py", line 98, in main
    summary = summarize(main_content)
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\91745\OneDrive\Desktop\Github_analyser\scripts\Summary\summarizer.py", line 11, in summarize
    chat_completion = client.chat.completions.create(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\91745\AppData\Local\Programs\Python\Python312\Lib\site-packages\openai\_utils\_utils.py", line 274, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\91745\AppData\Local\Programs\Python\Python312\Lib\site-packages\openai\resources\chat\completions.py", line 704, in create
    return self._post(
           ^^^^^^^^^^^
  File "C:\Users\91745\AppData\Local\Programs\Python\Python312\Lib\site-packages\openai\_base_client.py", line 1270, in post
    return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\91745\AppData\Local\Programs\Python\Python312\Lib\site-packages\openai\_base_client.py", line 947, in request
    return self._request(
           ^^^^^^^^^^^^^^
  File "C:\Users\91745\AppData\Local\Programs\Python\Python312\Lib\site-packages\openai\_base_client.py", line 1036, in _request
    return self._retry_request(
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\91745\AppData\Local\Programs\Python\Python312\Lib\site-packages\openai\_base_client.py", line 1085, in _retry_request   
    return self._request(
           ^^^^^^^^^^^^^^
  File "C:\Users\91745\AppData\Local\Programs\Python\Python312\Lib\site-packages\openai\_base_client.py", line 1036, in _request
    return self._retry_request(
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\91745\AppData\Local\Programs\Python\Python312\Lib\site-packages\openai\_base_client.py", line 1085, in _retry_request   
    return self._request(
           ^^^^^^^^^^^^^^
  File "C:\Users\91745\AppData\Local\Programs\Python\Python312\Lib\site-packages\openai\_base_client.py", line 1051, in _request
    raise self._make_status_error_from_response(err.response) from None
openai.InternalServerError: <html>
<head><title>504 Gateway Time-out</title></head>
<body>
<center><h1>504 Gateway Time-out</h1></center>
</body>
</html>
```

**Comment by juntao** [comment]
Hmm, this is a timeout error. I think you should set up your script to retry if 504 is encountered. 

Also, you can use the 

https://llama8b.gaia.domains/v1

Or

https://coder.gaia.domains/v1

which are less prone to timeouts.

**Comment by staru09** [comment]
I'll try these but what will be the MODEL_NAME of these endpoints?
Is it llama or something else?

**Comment by staru09** [comment]
> Also, you can use the
> 
> https://llama8b.gaia.domains/v1
> 
> Or
> 
> https://coder.gaia.domains/v1
> 
> which are less prone to timeouts.

I'll try these but what will be the MODEL_NAME of these endpoints?
Is it llama or something else?

**Comment by juntao** [comment]
https://llama8b.gaia.domains/v1 -- llama

https://llama3b.gaia.domains/v1 -- llama3b

**Comment by staru09** [comment]
Greetings sir
I have added links to standard repos, their corresponding final csv file (2 columns), and snapshots of their knowledge bases here in this [sheet](https://docs.google.com/spreadsheets/d/1_hRiQkVr9Dl2BLyjY87XjfReXUow0c9vKNA4nw_s34g/edit?usp=sharing) for easier access. Now some of the snapshots are yet to be added which I'll add there in a few days. 

What should I do after the knowledge base thing?

**Comment by staru09** [comment]
Greetings sir

> 2 Give the node a few tasks from the repo’s issue tracker and see if it can do this. We can use Cursor IDE for this so that we can easily run the generated code.

I created a node using this and deployed it using chatbot-ui but I was unable to access it using cursor ide. To deploy the node do I need to do as mentioned [here?](https://github.com/GaiaNet-AI/gaia-domain)
 
> 3 Set up a script to update the knowledge base on the node every hour.

for this I had planned to create a cronjob that will execute all the steps every hour
```
python3 github_parser.py
python3 summarizer.py

curl -sSf https://raw.githubusercontent.com/WasmEdge/WasmEdge/master/utils/install_v2.sh | bash -s
curl -LO https://huggingface.co/gaianet/Nomic-embed-text-v1.5-Embedding-GGUF/resolve/main/nomic-embed-text-v1.5.f16.gguf

mkdir -p qdrant_storage
mkdir -p qdrant_snapshots

nohup docker run -d -p 6333:6333 -p 6334:6334 \
    -v $(pwd)/qdrant_storage:/qdrant/storage:z \
    -v $(pwd)/qdrant_snapshots:/qdrant/snapshots:z \
    qdrant/qdrant

curl -X DELETE 'http://localhost:6333/collections/default'

curl -LO https://github.com/GaiaNet-AI/embedding-tools/raw/main/csv_embed/csv_embed.wasm
curl -LO https://huggingface.co/datasets/gaianet/paris/raw/main/paris.csv

wasmedge --dir .:. \
  --nn-preload embedding:GGML:AUTO:nomic-embed-text-v1.5.f16.gguf \
  csv_embed.wasm embedding default 768 paris.csv --ctx_size 8192

curl -X POST 'http://localhost:6333/collections/default/snapshots'

sudo tar czvf my.snapshot.tar.gz my.snapshot
```
using this I can create an updated snapshot but if we have to create a deployable node then I think [these](https://docs.gaianet.ai/node-guide/quick-start) steps can be followed.

**Comment by juntao** [comment]
Thank you @staru09 Would love to know the issues you encountered when [connecting your gaia node to cursor](https://docs.gaianet.ai/user-guide/apps/cursor). Can you elaborate?

Now, we we draw close to the end of the project, lets create a complete product to demonstrate everything together. I would like to create a coding assistant for the popular Python framework [Gradio](https://www.gradio.app/)

- [ ] Script to automatically generate QA CSV for all MD files (English only) under: https://github.com/gradio-app/gradio/tree/main/guides
- [ ] Script to automatically generate QA CSV for all Python files under: https://github.com/gradio-app/gradio/tree/main/demo -- I am thinking that you may need to handwrite 2-3 Qs for each `run.py` file since it probably hard for the LLM to figure out what it does by just looking at the code.
- [ ] Generated Qdrant at a specific time (you choose the time)
- [ ] An llm.txt document to teach people the basics of a Gradio project. 

Please submit a complete repo that contains

* All scripts
* Documents on how to run them step by step (starting from how to start a qdrant server)

Thanks.

**Comment by staru09** [comment]
> Thank you @staru09 Would love to know the issues you encountered when [connecting your gaia node to cursor](https://docs.gaianet.ai/user-guide/apps/cursor). Can you elaborate?

Sir I ran this [node](https://github.com/GaiaNet-AI/node-configs/tree/main/qwen-2.5-coder-1.5b-instruct)  and when I try to add it to the cursor IDE it gave me an invalid api error.

```qwen-2.5-coder-1.5b``` was the model name 
```https://0x8c66498c8124d55a7330e832c56a44223fb9622f.us.gaianet.network/v1``` was the api endpoint
but it gave me this error 
![node](https://github.com/user-attachments/assets/33a32dbc-d1c3-435f-ab92-a9d5c9bc5b78)

these are the node config and cursor settings 
![Screenshot from 2024-11-24 23-31-38](https://github.com/user-attachments/assets/09a29d0f-3cc9-4422-9899-68d762ab867e)

![Screenshot from 2024-11-24 23-31-09](https://github.com/user-attachments/assets/5f4313dc-bcf5-4ff4-abb6-31fa16fd91b6)

**Comment by juntao** [comment]
I tried 

```
curl https://0x8c66498c8124d55a7330e832c56a44223fb9622f.us.gaianet.network/v1/models
```

But it says node not started. I guess that is because it is running on your laptop and you have stopped it (or simply closed the laptop). 

Regardless, lets complete tasks for the gradio project. Thanks.

**Comment by staru09** [comment]
Greetings sir 
This [repo](https://github.com/staru09/Gradio_bot) contains all the scripts needed to create a knowledge base from Gradio. 
I have added all the steps in the documentation to run the scripts and create a knowledge base.

**Comment by staru09** [comment]
Greetings sir
I have submitted all the scripts needed to create a self updating knowledge
base from a github repo. The script will run automatically at 12 every
night. Now I am only left with creating a demo bot for Gradio and a llm.txt
file for the project which I will complete by the end of this week because
I'll be having exam till Wednesday in our university. I sincerely apologize
for this slight delay and request your understanding.

Thank You
Aru

**Comment by juntao** [comment]
Are you running it inside Docker? Docker typically only allocate 2-4GB of RAM for the container, which is far too small. You can try a tiny model like the Qwen 0.5B

https://github.com/GaiaNet-AI/node-configs/tree/main/qwen2-0.5b-instruct

**Comment by juntao** [comment]
Also, what is the URL format should I use?

https://github.com/owner/repo

or 

https://github.com/owner/repo/tree/main

the comments in the script seems to suggest the latter.

**Comment by juntao** [comment]
What’s your system? OS, CPU, RAM etc

**Comment by staru09** [comment]
> Also, what is the URL format should I use?
> 
> https://github.com/owner/repo
> 
> or
> 
> https://github.com/owner/repo/tree/main
> 
> the comments in the script seems to suggest the latter.

Sir, you can use the URL in any format. The script basically checks if it has to fetch from the base repo or from any specific folder inside the base repo. 

```
def get_github_contents(repo_url):
    print("Starting to fetch GitHub repository contents...")
    parts = repo_url.rstrip('/').split('/')
    
    if len(parts) < 5 or parts[2] != "github.com":
        print("Error: Invalid GitHub URL. Please check the format.")
        raise ValueError("Invalid GitHub URL. Ensure the URL is in the format: https://github.com/user/repo/tree/branch/path")
    
    user = parts[3]
    repo = parts[4]
    
    if "tree" in parts:
        branch = parts[6]
        subpath = '/'.join(parts[7:]) if len(parts) > 7 else ''
        api_url = f"https://api.github.com/repos/{user}/{repo}/contents/{subpath}?ref={branch}"
    else:
        api_url = f"https://api.github.com/repos/{user}/{repo}/contents/"
```

**Comment by staru09** [comment]
> What’s your system? OS, CPU, RAM etc

Sir, it has Ubuntu 24.04 LTS as OS, Intel Xeon as CPU and the system has around 64 GB RAM along with a 24 GB GPU (NVIDIA A5500)

I have attached the PC specs in the previous screenshot.

**Comment by juntao** [comment]
Can you add a feature to support skipping certain folders matching a regex expression? For example, we might want to skip all files under the `vendor` folder of our repo. 

> > Also, what is the URL format should I use?
> 
> > 
> 
> > https://github.com/owner/repo
> 
> > 
> 
> > or
> 
> > 
> 
> > https://github.com/owner/repo/tree/main
> 
> > 
> 
> > the comments in the script seems to suggest the latter.
> 
> 
> 
> Sir, you can use the URL in any format. The script basically checks if it has to fetch from the base repo or from any specific folder inside the base repo. 
> 
> 
> 
> ```
> 
> def get_github_contents(repo_url):
> 
>     print("Starting to fetch GitHub repository contents...")
> 
>     parts = repo_url.rstrip('/').split('/')
> 
>     
> 
>     if len(parts) < 5 or parts[2] != "github.com":
> 
>         print("Error: Invalid GitHub URL. Please check the format.")
> 
>         raise ValueError("Invalid GitHub URL. Ensure the URL is in the format: https://github.com/user/repo/tree/branch/path")
> 
>     
> 
>     user = parts[3]
> 
>     repo = parts[4]
> 
>     
> 
>     if "tree" in parts:
> 
>         branch = parts[6]
> 
>         subpath = '/'.join(parts[7:]) if len(parts) > 7 else ''
> 
>         api_url = f"https://api.github.com/repos/{user}/{repo}/contents/{subpath}?ref={branch}"
> 
>     else:
> 
>         api_url = f"https://api.github.com/repos/{user}/{repo}/contents/"
> 
> ```

**Comment by juntao** [comment]
> I reinstalled the docker and then tried running the node. The node still isn't starting but its also not showing the earlier dump message. 
> 
> ![image](https://github.com/user-attachments/assets/0f22c7a7-7637-499e-b952-05f89ce3be32)
> 
> 

Do a 

```
gaianet stop

```

And then


```
gaianet start

```

Do not use `--local-only`

**Comment by staru09** [comment]
> Can you add a feature to support skipping certain folders matching a regex expression? For example, we might want to skip all files under the `vendor` folder of our repo.

Sir this is how I was skipping the folders at [gradio](https://github.com/gradio-app/gradio/tree/main/guides) 
Since I have to fetch only English-related files skipping 'cn' folder.

```
md_file.py

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python script.py <repo_url> <output_csv_path>")
        sys.exit(1)

    repo_url = sys.argv[1]
    output_path = sys.argv[2]

    exclude_folders = ['cn']

    contents = get_github_contents(repo_url)
    paths = process_contents(contents, exclude_folders=exclude_folders)

    write_to_csv(paths, output_path)
    print(f"CSV file '{output_path}' generated successfully.")
    
```

**Comment by juntao** [comment]
Can we incorporate this feature into `md_file` and `python_file` scripts and make it an option like the following?

```
python md_file.py <repo_link> <path_to_output.csv> --exclude "regex_for_excluded_file_or_folders"
```

**Comment by staru09** [comment]
Greetings sir
I just wanted to ask if the project is completed or do I need to add
something more to it?

Thanks
Aru

**Comment by juntao** [comment]
Thanks. I am a little confused now ... What is the difference between

https://github.com/staru09/Gradio_bot
 
and

https://github.com/staru09/Github_analyser

The script names are different. If Github_analyser is the main repo, we should make Gradio_bot an "example project" inside the main repo. Make sure that they use the same script files to generate the knowledge base.

Also, we need to contribute this repo to the WasmEdge project. I forked the Github_analyser repo at

https://github.com/LlamaEdge/github_repo_to_llm

Can you delete your upstream repo and then make the above changes (e.g., make a gradio an example) in the new repo through a PR? Thanks.

**Comment by juntao** [comment]
Please make sure that you document the `--exclude` flag in the README for the appropriate scripts. Thanks.

**Comment by juntao** [comment]
Sorry for the delay. We have holidays in the US. There are only a few items
left.

https://github.com/WasmEdge/WasmEdge/issues/3725#issuecomment-2508777146

We will use these scripts to create a knowledge bot for the CRI-O project.
Would you like to come to our community meeting (virtual over zoom) on Jan
7th to present this work? Thanks!

cheers
Michael
---
Michael Yuan, PhD
Star WasmEdge <https://github.com/WasmEdge/WasmEdge> on GitHub!

[image: Chat, App, Social, Game, gamer, gamers, Discord Icon]
<https://discord.gg/EEWsdHcNW3>  [image: youtube icon download 32x32 -
curved] <https://www.youtube.com/c/SecondStateInc>  [image: twitter icon
download 32x32 - curved] <https://twitter.com/realwasmedge>
<https://wasmedge.org/book/en/>

**Comment by staru09** [comment]
> Thanks. I am a little confused now ... What is the difference between
> 
> https://github.com/staru09/Gradio_bot
> 
> and
> 
> https://github.com/staru09/Github_analyser
> 
I am so sorry for this confusion. The main repo is Github_analyser only and Gradio bot is more like an example repo.
I'll add it the main repo and update all the scripts with the ```--exclude flag```.

**Comment by staru09** [comment]
Greetings Sir,
I have updated the scripts in the main repo with ```--exclude``` flag and have added a link to the Gradio Bot in the main repo.

**Comment by juntao** [comment]
Hmm, I do not see a PR here?

https://github.com/LlamaEdge/github_repo_to_llm

**Comment by juntao** [comment]
And I do not see the gradio_bot example in the main repo?

**Comment by staru09** [comment]
> Hmm, I do not see a PR here?
> 
> https://github.com/LlamaEdge/github_repo_to_llm

I am so sorry sir I have added the link in the main documentation becuase there was an issue related to nested repositories in the repo. I'll try to push a fix in some time.

**Comment by juntao** [comment]
I would like the example to be just code under the `examples` directory instead of another repo in your personal account. Also,

1 The example uses scripts that are NOT in the main repo. That is not really an example but another project. The example should demonstrate how to use the scripts in the main repo.

2 Please do all these in the repo in the LlamaEdge org. It would not count as project contribution in your personal repo.

Thanks.

**Comment by hydai** [comment]
Hi @juntao and @staru09 
Just want to check if this issue is resolved or not.

**Comment by staru09** [comment]
I think this is over but we can add something agentic like MCPs, some reranking methods etc.
wdyt @juntao ?

**Comment by hydai** [comment]
Since it's over, closing this issue.
